File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/3-H5支付/PartnerH5Prepay.go
package main

import (
	"bytes"
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

func main() {
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",
		"YOUR_CERT_SERIAL_NO",
		"/path/to/apiclient_key.pem",
		"YOUR_PUB_KEY_ID",
		"/path/to/wxp_pub.pem",
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &PartnerApiv3PartnerH5PrepayRequest{
		SpAppid:     wxpay_utility.String("wxYOUR_APPID"),
		SpMchid:     wxpay_utility.String("1230000109"),
		SubAppid:    wxpay_utility.String("wxd678efh567hg6999"),
		SubMchid:    wxpay_utility.String("1900000109"),
		Description: wxpay_utility.String("Image形象店-深圳腾大-QQ公仔"),
		OutTradeNo:  wxpay_utility.String("1217752501201407033233368018"),
		TimeExpire:  wxpay_utility.Time(time.Now()),
		Attach:      wxpay_utility.String("自定义数据说明"),
		NotifyUrl:   wxpay_utility.String(" https://www.weixin.qq.com/wxpay/pay.php"),
		GoodsTag:    wxpay_utility.String("WXG"),
		SettleInfo: &PartnerSettleInfo{
			ProfitSharing: wxpay_utility.Bool(true),
		},
		SupportFapiao: wxpay_utility.Bool(false),
		Amount: &CommonAmountInfo{
			Total:    wxpay_utility.Int64(100),
			Currency: wxpay_utility.String("CNY"),
		},
		Detail: &CouponInfo{
			CostPrice: wxpay_utility.Int64(608800),
			InvoiceId: wxpay_utility.String("微信123"),
			GoodsDetail: []GoodsDetail{
				{
					MerchantGoodsId:  wxpay_utility.String("1246464644"),
					WechatpayGoodsId: wxpay_utility.String("1001"),
					GoodsName:        wxpay_utility.String("iPhoneX 256G"),
					Quantity:         wxpay_utility.Int64(1),
					UnitPrice:        wxpay_utility.Int64(528800),
				},
			},
		},
		SceneInfo: &H5ReqSceneInfo{
			PayerClientIp: wxpay_utility.String("14.23.150.211"),
			DeviceId:      wxpay_utility.String("013467007045764"),
			StoreInfo: &StoreInfo{
				Id:       wxpay_utility.String("0001"),
				Name:     wxpay_utility.String("腾讯大厦分店"),
				AreaCode: wxpay_utility.String("440305"),
				Address:  wxpay_utility.String("广东省深圳市南山区科技中一道10000号"),
			},
			H5Info: &H5Info{
				Type:        wxpay_utility.String("iOS"),
				AppName:     wxpay_utility.String("王者荣耀"),
				AppUrl:      wxpay_utility.String("https://pay.qq.com"),
				BundleId:    wxpay_utility.String("com.tencent.wzryiOS"),
				PackageName: wxpay_utility.String("com.tencent.tmgp.sgame"),
			},
		},
	}

	response, err := PartnerH5Prepay(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		return
	}
	fmt.Printf("请求成功: %+v\n", response)
}

func PartnerH5Prepay(config *wxpay_utility.MchConfig, request *PartnerApiv3PartnerH5PrepayRequest) (response *PartnerApiv3PartnerH5PrepayResponse, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "POST"
		path   = "/v3/pay/partner/transactions/h5"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqBody, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}
	httpRequest, err := http.NewRequest(method, reqUrl.String(), bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	httpRequest.Header.Set("Content-Type", "application/json")
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), reqBody)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		err = wxpay_utility.ValidateResponse(config.WechatPayPublicKeyId(), config.WechatPayPublicKey(), &httpResponse.Header, respBody)
		if err != nil {
			return nil, err
		}
		response := &PartnerApiv3PartnerH5PrepayResponse{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}
		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(httpResponse.StatusCode, httpResponse.Header, respBody)
	}
}

type PartnerApiv3PartnerH5PrepayRequest struct {
	SpAppid       *string            `json:"sp_appid,omitempty"`
	SpMchid       *string            `json:"sp_mchid,omitempty"`
	SubAppid      *string            `json:"sub_appid,omitempty"`
	SubMchid      *string            `json:"sub_mchid,omitempty"`
	Description   *string            `json:"description,omitempty"`
	OutTradeNo    *string            `json:"out_trade_no,omitempty"`
	TimeExpire    *time.Time         `json:"time_expire,omitempty"`
	Attach        *string            `json:"attach,omitempty"`
	NotifyUrl     *string            `json:"notify_url,omitempty"`
	GoodsTag      *string            `json:"goods_tag,omitempty"`
	SettleInfo    *PartnerSettleInfo `json:"settle_info,omitempty"`
	SupportFapiao *bool              `json:"support_fapiao,omitempty"`
	Amount        *CommonAmountInfo  `json:"amount,omitempty"`
	Detail        *CouponInfo        `json:"detail,omitempty"`
	SceneInfo     *H5ReqSceneInfo    `json:"scene_info,omitempty"`
}

type PartnerApiv3PartnerH5PrepayResponse struct {
	H5Url *string `json:"h5_url,omitempty"`
}

type PartnerSettleInfo struct {
	ProfitSharing *bool `json:"profit_sharing,omitempty"`
}

type CommonAmountInfo struct {
	Total    *int64  `json:"total,omitempty"`
	Currency *string `json:"currency,omitempty"`
}

type CouponInfo struct {
	CostPrice   *int64        `json:"cost_price,omitempty"`
	InvoiceId   *string       `json:"invoice_id,omitempty"`
	GoodsDetail []GoodsDetail `json:"goods_detail,omitempty"`
}

type H5ReqSceneInfo struct {
	PayerClientIp *string    `json:"payer_client_ip,omitempty"`
	DeviceId      *string    `json:"device_id,omitempty"`
	StoreInfo     *StoreInfo `json:"store_info,omitempty"`
	H5Info        *H5Info    `json:"h5_info,omitempty"`
}

type GoodsDetail struct {
	MerchantGoodsId  *string `json:"merchant_goods_id,omitempty"`
	WechatpayGoodsId *string `json:"wechatpay_goods_id,omitempty"`
	GoodsName        *string `json:"goods_name,omitempty"`
	Quantity         *int64  `json:"quantity,omitempty"`
	UnitPrice        *int64  `json:"unit_price,omitempty"`
}

type StoreInfo struct {
	Id       *string `json:"id,omitempty"`
	Name     *string `json:"name,omitempty"`
	AreaCode *string `json:"area_code,omitempty"`
	Address  *string `json:"address,omitempty"`
}

type H5Info struct {
	Type        *string `json:"type,omitempty"`
	AppName     *string `json:"app_name,omitempty"`
	AppUrl      *string `json:"app_url,omitempty"`
	BundleId    *string `json:"bundle_id,omitempty"`
	PackageName *string `json:"package_name,omitempty"`
}

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Java/3-H5支付/PartnerH5Prepay.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/partner/4014985777

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * H5下单
 */
public class PartnerH5Prepay {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "POST";
  private static String PATH = "/v3/pay/partner/transactions/h5";

  public static void main(String[] args) {
    // TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/partner/4013080340
    PartnerH5Prepay client = new PartnerH5Prepay(
      "YOUR_MCHID",
      "YOUR_CERT_SERIAL_NO",
      "/path/to/apiclient_key.pem",
      "YOUR_PUB_KEY_ID",
      "/path/to/wxp_pub.pem"
    );

    PartnerAPIv3PartnerH5PrepayRequest request = new PartnerAPIv3PartnerH5PrepayRequest();
    request.spAppid = "wxYOUR_APPID";
    request.spMchid = "1230000109";
    request.subAppid = "wxd678efh567hg6999";
    request.subMchid = "1900000109";
    request.description = "Image形象店-深圳腾大-QQ公仔";
    request.outTradeNo = "1217752501201407033233368018";
    request.timeExpire = "2018-06-08T10:34:56+08:00";
    request.attach = "自定义数据说明";
    request.notifyUrl = " https://www.weixin.qq.com/wxpay/pay.php";
    request.goodsTag = "WXG";
    request.settleInfo = new PartnerSettleInfo();
    request.settleInfo.profitSharing = true;
    request.supportFapiao = false;
    request.amount = new CommonAmountInfo();
    request.amount.total = 100L;
    request.amount.currency = "CNY";
    request.detail = new CouponInfo();
    request.detail.costPrice = 608800L;
    request.detail.invoiceId = "微信123";
    request.detail.goodsDetail = new ArrayList<>();
    {
      GoodsDetail goodsDetailItem0 = new GoodsDetail();
      goodsDetailItem0.merchantGoodsId = "1246464644";
      goodsDetailItem0.wechatpayGoodsId = "1001";
      goodsDetailItem0.goodsName = "iPhoneX 256G";
      goodsDetailItem0.quantity = 1L;
      goodsDetailItem0.unitPrice = 528800L;
      request.detail.goodsDetail.add(goodsDetailItem0);
    };
    request.sceneInfo = new H5ReqSceneInfo();
    request.sceneInfo.payerClientIp = "14.23.150.211";
    request.sceneInfo.deviceId = "013467007045764";
    request.sceneInfo.storeInfo = new StoreInfo();
    request.sceneInfo.storeInfo.id = "0001";
    request.sceneInfo.storeInfo.name = "腾讯大厦分店";
    request.sceneInfo.storeInfo.areaCode = "440305";
    request.sceneInfo.storeInfo.address = "广东省深圳市南山区科技中一道10000号";
    request.sceneInfo.h5Info = new H5Info();
    request.sceneInfo.h5Info.type = "iOS";
    request.sceneInfo.h5Info.appName = "王者荣耀";
    request.sceneInfo.h5Info.appUrl = "https://pay.qq.com";
    request.sceneInfo.h5Info.bundleId = "com.tencent.wzryiOS";
    request.sceneInfo.h5Info.packageName = "com.tencent.tmgp.sgame";
    try {
      PartnerAPIv3PartnerH5PrepayResponse response = client.run(request);
        System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
        e.printStackTrace();
    }
  }

  public PartnerAPIv3PartnerH5PrepayResponse run(PartnerAPIv3PartnerH5PrepayRequest request) {
    String uri = PATH;
    String reqBody = WXPayUtility.toJson(request);

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo,privateKey, METHOD, uri, reqBody));
    reqBuilder.addHeader("Content-Type", "application/json");
    RequestBody requestBody = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), reqBody);
    reqBuilder.method(METHOD, requestBody);
    Request httpRequest = reqBuilder.build();

    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);
        return WXPayUtility.fromJson(respBody, PartnerAPIv3PartnerH5PrepayResponse.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public PartnerH5Prepay(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class PartnerAPIv3PartnerH5PrepayRequest {
    @SerializedName("sp_appid") public String spAppid;
    @SerializedName("sp_mchid") public String spMchid;
    @SerializedName("sub_appid") public String subAppid;
    @SerializedName("sub_mchid") public String subMchid;
    @SerializedName("description") public String description;
    @SerializedName("out_trade_no") public String outTradeNo;
    @SerializedName("time_expire") public String timeExpire;
    @SerializedName("attach") public String attach;
    @SerializedName("notify_url") public String notifyUrl;
    @SerializedName("goods_tag") public String goodsTag;
    @SerializedName("settle_info") public PartnerSettleInfo settleInfo;
    @SerializedName("support_fapiao") public Boolean supportFapiao;
    @SerializedName("amount") public CommonAmountInfo amount;
    @SerializedName("detail") public CouponInfo detail;
    @SerializedName("scene_info") public H5ReqSceneInfo sceneInfo;
  }

  public static class PartnerAPIv3PartnerH5PrepayResponse {
    @SerializedName("h5_url") public String h5Url;
  }

  public static class PartnerSettleInfo {
    @SerializedName("profit_sharing") public Boolean profitSharing;
  }

  public static class CommonAmountInfo {
    @SerializedName("total") public Long total;
    @SerializedName("currency") public String currency;
  }

  public static class CouponInfo {
    @SerializedName("cost_price") public Long costPrice;
    @SerializedName("invoice_id") public String invoiceId;
    @SerializedName("goods_detail") public List<GoodsDetail> goodsDetail;
  }

  public static class H5ReqSceneInfo {
    @SerializedName("payer_client_ip") public String payerClientIp;
    @SerializedName("device_id") public String deviceId;
    @SerializedName("store_info") public StoreInfo storeInfo;
    @SerializedName("h5_info") public H5Info h5Info;
  }

  public static class GoodsDetail {
    @SerializedName("merchant_goods_id") public String merchantGoodsId;
    @SerializedName("wechatpay_goods_id") public String wechatpayGoodsId;
    @SerializedName("goods_name") public String goodsName;
    @SerializedName("quantity") public Long quantity;
    @SerializedName("unit_price") public Long unitPrice;
  }

  public static class StoreInfo {
    @SerializedName("id") public String id;
    @SerializedName("name") public String name;
    @SerializedName("area_code") public String areaCode;
    @SerializedName("address") public String address;
  }

  public static class H5Info {
    @SerializedName("type") public String type;
    @SerializedName("app_name") public String appName;
    @SerializedName("app_url") public String appUrl;
    @SerializedName("bundle_id") public String bundleId;
    @SerializedName("package_name") public String packageName;
  }

}
